File: src/com/anathema_roguelike/entities/characters/perks/actions/targetingstrategies/ConalAreaOfEffect.java
package com.anathema_roguelike.entities.characters.perks.actions.targetingstrategies;

import com.anathema_roguelike.entities.characters.perks.actions.targetingstrategies.shapes.Cone;
import com.anathema_roguelike.entities.characters.perks.actions.targetingstrategies.shapes.Shape;
import com.anathema_roguelike.main.utilities.position.Direction;

public abstract class ConalAreaOfEffect<TargetType extends Targetable, OriginType extends Targetable> extends AreaOfEffect<TargetType, OriginType> {

	private AdditiveCalculation radius;
	
	@SafeVarargs
	public ConalAreaOfEffect(Class<TargetType> targetType, AdditiveCalculation radius, TargetConstraint<TargetType, OriginType> ...constraints) {
		super(targetType, constraints);
		
		this.radius = radius;
	}

	@Override
	protected Shape getShape(OriginType target) {
		return new Cone(getOrigin(), radius.get(), Direction.of(getOrigin(), target));
	}
	
	protected abstract OriginType getOrigin();
}

File: src/com/anathema_roguelike/entities/characters/perks/actions/targetingstrategies/ConalAreaOfEffect.scala


package com.anathema_roguelike
package entities.characters.perks.actions.targetingstrategies

import com.anathema_roguelike.entities.characters.perks.actions.targetingstrategies.constraints.TargetConstraint
import com.anathema_roguelike.entities.characters.perks.actions.targetingstrategies.shapes.{Cone, Shape}
import com.anathema_roguelike.main.utilities.position.Direction
import com.anathema_roguelike.stats.effects.AdditiveCalculation

import scala.reflect.ClassTag
import scala.reflect.runtime.universe._

abstract class ConalAreaOfEffect[TargetType <: Targetable : TypeTag : ClassTag, ArgType <: Targetable : TypeTag](
    var radius: AdditiveCalculation,
    constraints: TargetConstraint[TargetType, ArgType]*
  ) extends AreaOfEffect[TargetType, ArgType](constraints:_*) {

  override protected def getShape (target: ArgType): Shape = {
    new Cone(getOrigin, radius(), Direction.of(getOrigin, target) )
  }

  protected def getOrigin: ArgType
}
